File: bindings/javascript/packages/core/src/noise.ts
/**
 * Noise channel bindings.
 *
 * Apply realistic decoherence channels to a {@link QuantumState}:
 * depolarising, amplitude-damping, phase-damping, pure-dephasing,
 * bit-flip, phase-flip, bit-phase-flip, thermal relaxation, plus
 * classical readout error and the composite per-device noise model
 * (`noise_model_t` + `noise_apply_model`).
 *
 * All channels are Monte-Carlo trajectory unravellings of the
 * Kraus-operator CPTP map: each call realises ONE Kraus branch selected
 * by a uniform-[0, 1) random draw, exactly as the C engine defines it
 * (`src/quantum/noise.{c,h}`).  Averaging the reduced density operator
 * over many trajectories recovers the channel.  The state is mutated in
 * place; norm is preserved up to floating-point precision.
 *
 * Every C channel takes the random draw(s) as trailing argument(s).
 * The JS layer supplies them: by default from a crypto-backed uniform
 * generator, or from the optional trailing parameter so tests and
 * reproducible pipelines can inject their own stream.  Mirrors Python
 * `moonlab.noise` (seedable numpy Generator) and Rust `moonlab::noise`
 * (caller-supplied `random_value`).
 *
 * WASM exports declared in
 * `bindings/javascript/packages/core/emscripten/exports.txt`.
 *
 * @since v0.2.1 (channel surface); v1.2.0 (correct random-draw
 *   plumbing, readout error, DeviceNoiseModel)
 */

import type { QuantumState } from './quantum-state';
import { getModule } from './wasm-loader';

interface NoiseModule {
  _noise_depolarizing_single: (s: number, q: number, p: number, rv: number) => void;
  _noise_depolarizing_two_qubit: (s: number, q1: number, q2: number, p: number, rv: number) => void;
  _noise_amplitude_damping: (s: number, q: number, gamma: number, rv: number) => void;
  _noise_phase_damping: (s: number, q: number, gamma: number, rv: number) => void;
  _noise_pure_dephasing: (s: number, q: number, sigma: number, randomPhase: number) => void;
  _noise_bit_flip: (s: number, q: number, p: number, rv: number) => void;
  _noise_phase_flip: (s: number, q: number, p: number, rv: number) => void;
  _noise_bit_phase_flip: (s: number, q: number, p: number, rv: number) => void;
  _noise_thermal_relaxation: (
    s: number, q: number, t1: number, t2: number, t: number, rvsPtr: number,
  ) => void;
  _noise_readout_error: (outcome: number, e01: number, e10: number, rv: number) => number;
  _noise_model_create: () => number;
  _noise_model_destroy: (m: number) => void;
  _noise_model_create_realistic: (
    t1Us: number, t2Us: number, gateError: number, readoutError: number,
  ) => number;
  _noise_model_set_depolarizing: (m: number, rate: number) => void;
  _noise_model_set_amplitude_damping: (m: number, rate: number) => void;
  _noise_model_set_phase_damping: (m: number, rate: number) => void;
  _noise_model_set_thermal: (m: number, t1: number, t2: number) => void;
  _noise_model_set_gate_time: (m: number, t: number) => void;
  _noise_model_set_readout_error: (m: number, e01: number, e10: number) => void;
  _noise_model_set_enabled: (m: number, enabled: number) => void;
  _noise_apply_model: (s: number, q: number, m: number, rvsPtr: number) => void;
  _noise_apply_model_two_qubit: (
    s: number, q1: number, q2: number, m: number, rvsPtr: number,
  ) => void;
  _malloc: (n: number) => number;
  _free: (p: number) => void;
  HEAPF64: Float64Array;
}

function modAndPtr(state: QuantumState): { mod: NoiseModule; ptr: number } {
  const mod = state._internal_module() as unknown as NoiseModule;
  const ptr = state._internal_state_pointer();
  return { mod, ptr };
}

/** Uniform-[0, 1) source used when the caller does not supply a draw. */
export type RandomSource = () => number;

/** Crypto-backed uniform-[0, 1) draw (53-bit mantissa). */
function cryptoUniform(): number {
  const buf = new Uint32Array(2);
  globalThis.crypto.getRandomValues(buf);
  // 26 + 27 bits -> uniform double in [0, 1) with full 53-bit precision.
  return ((buf[0] >>> 6) * 134217728 + (buf[1] >>> 5)) / 9007199254740992;
}

const defaultRng: RandomSource = cryptoUniform;

function draw(rv: number | undefined): number {
  return rv === undefined ? defaultRng() : rv;
}

/** Standard-normal draw (Box-Muller over the default uniform source). */
function normalDraw(): number {
  let u = defaultRng();
  if (u <= 0) u = Number.MIN_VALUE;
  const v = defaultRng();
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

/** Single-qubit depolarising channel with probability `p`.
 *  `randomValue` is the uniform-[0, 1) branch selector; omit it to use
 *  the crypto-backed default. */
export function depolarizing(
  state: QuantumState, qubit: number, p: number, randomValue?: number,
): void {
  const { mod, ptr } = modAndPtr(state);
  mod._noise_depolarizing_single(ptr, qubit, p, draw(randomValue));
}

/** Two-qubit depolarising channel with probability `p`. */
export function depolarizing2(
  state: QuantumState, qubit1: number, qubit2: number, p: number,
  randomValue?: number,
): void {
  const { mod, ptr } = modAndPtr(state);
  mod._noise_depolarizing_two_qubit(ptr, qubit1, qubit2, p, draw(randomValue));
}

/** Amplitude-damping channel (T1 relaxation) with rate `gamma` in [0, 1]. */
export function amplitudeDamping(
  state: QuantumState, qubit: number, gamma: number, randomValue?: number,
): void {
  const { mod, ptr } = modAndPtr(state);
  mod._noise_amplitude_damping(ptr, qubit, gamma, draw(randomValue));
}

/** Phase-damping channel with rate `gamma` in [0, 1]. */
export function phaseDamping(
  state: QuantumState, qubit: number, gamma: number, randomValue?: number,
): void {
  const { mod, ptr } = modAndPtr(state);
  mod._noise_phase_damping(ptr, qubit, gamma, draw(randomValue));
}

/** Gaussian pure-dephasing: multiplies the qubit's |1> amplitudes by
 *  `exp(i * sigma * randomPhase)`.  `randomPhase` should be a standard
 *  normal draw; omit it to use a Box-Muller draw over the default
 *  source, making the accumulated phase variance `sigma^2` per call. */
export function pureDephasing(
  state: QuantumState, qubit: number, sigma: number, randomPhase?: number,
): void {
  const { mod, ptr } = modAndPtr(state);
  mod._noise_pure_dephasing(
    ptr, qubit, sigma, randomPhase === undefined ? normalDraw() : randomPhase,
  );
}

/** Bit-flip channel with probability `p`. */
export function bitFlip(
  state: QuantumState, qubit: number, p: number, randomValue?: number,
): void {
  const { mod, ptr } = modAndPtr(state);
  mod._noise_bit_flip(ptr, qubit, p, draw(randomValue));
}

/** Phase-flip channel with probability `p`. */
export function phaseFlip(
  state: QuantumState, qubit: number, p: number, randomValue?: number,
): void {
  const { mod, ptr } = modAndPtr(state);
  mod._noise_phase_flip(ptr, qubit, p, draw(randomValue));
}

/** Bit-phase-flip channel with probability `p`. */
export function bitPhaseFlip(
  state: QuantumState, qubit: number, p: number, randomValue?: number,
): void {
  const { mod, ptr } = modAndPtr(state);
  mod._noise_bit_phase_flip(ptr, qubit, p, draw(randomValue));
}

/** Thermal relaxation with T1, T2 timescales over duration `t`.  The C
 *  channel consumes two uniform draws (T1 decay branch, T2 dephasing
 *  branch); pass `randomValues` to inject them. */
export function thermalRelaxation(
  state: QuantumState,
  qubit: number,
  t1: number,
  t2: number,
  duration: number,
  randomValues?: readonly [number, number],
): void {
  const { mod, ptr } = modAndPtr(state);
  const rvs = randomValues ?? [defaultRng(), defaultRng()];
  const rvsPtr = mod._malloc(16);
  try {
    mod.HEAPF64[rvsPtr >> 3] = rvs[0];
    mod.HEAPF64[(rvsPtr >> 3) + 1] = rvs[1];
    mod._noise_thermal_relaxation(ptr, qubit, t1, t2, duration, rvsPtr);
  } finally {
    mod._free(rvsPtr);
  }
}

/** Classical readout error: flips a measured `outcome` (0 or 1) with
 *  probability `error0to1` (reported 1 given true 0) respectively
 *  `error1to0` (reported 0 given true 1).  Pure post-processing -- the
 *  quantum state is untouched.  Returns the reported outcome. */
export async function readoutError(
  outcome: number, error0to1: number, error1to0: number, randomValue?: number,
): Promise<number> {
  const mod = (await getModule()) as unknown as NoiseModule;
  return mod._noise_readout_error(outcome, error0to1, error1to0, draw(randomValue));
}

/** Options for {@link DeviceNoiseModel.create}. */
export interface DeviceNoiseModelOptions {
  /** Single-qubit depolarising rate applied after each gate. */
  depolarizingRate?: number;
  /** Two-qubit depolarising rate (via `noise_model_create_realistic`'s
   *  gate-error scaling this is 10x the single-qubit rate when the
   *  realistic constructor is used; here it is whatever the C model
   *  holds -- set the single-qubit channels individually). */
  amplitudeDampingRate?: number;
  /** Phase-damping rate. */
  phaseDampingRate?: number;
  /** T1 relaxation time (same time unit as `gateTime`). */
  t1?: number;
  /** T2 dephasing time (same time unit as `gateTime`). */
  t2?: number;
  /** Gate duration driving the thermal-relaxation contribution. */
  gateTime?: number;
  /** Readout error probabilities [P(1|0), P(0|1)]. */
  readoutError?: readonly [number, number];
  /** Uniform-[0, 1) source for the per-application draws (defaults to
   *  the crypto-backed generator). */
  rng?: RandomSource;
}

/**
 * Composite per-device noise profile: depolarising + damping + thermal
 * relaxation applied together after each gate, mirroring Python
 * `moonlab.noise.DeviceNoiseModel`.  Wraps the C `noise_model_t` and
 * `noise_apply_model` / `noise_apply_model_two_qubit` entries.
 *
 * @example
 * ```typescript
 * const model = await DeviceNoiseModel.create({
 *   depolarizingRate: 1e-3, t1: 50e-6, t2: 30e-6, gateTime: 20e-9,
 * });
 * const state = await QuantumState.create({ numQubits: 2 });
 * state.h(0);
 * model.applySingle(state, 0);    // one noise trajectory after the gate
 * model.dispose();
 * ```
 */
export class DeviceNoiseModel {
  private ptr: number;
  private mod: NoiseModule;
  private rng: RandomSource;
  private freed = false;

  private constructor(ptr: number, mod: NoiseModule, rng: RandomSource) {
    this.ptr = ptr;
    this.mod = mod;
    this.rng = rng;
  }

  /** Build a model from explicit per-channel rates. */
  static async create(options: DeviceNoiseModelOptions = {}): Promise<DeviceNoiseModel> {
    const mod = (await getModule()) as unknown as NoiseModule;
    const ptr = mod._noise_model_create();
    if (!ptr) throw new Error('noise_model_create returned NULL');
    const m = new DeviceNoiseModel(ptr, mod, options.rng ?? defaultRng);
    if (options.depolarizingRate !== undefined) {
      mod._noise_model_set_depolarizing(ptr, options.depolarizingRate);
    }
    if (options.amplitudeDampingRate !== undefined) {
      mod._noise_model_set_amplitude_damping(ptr, options.amplitudeDampingRate);
    }
    if (options.phaseDampingRate !== undefined) {
      mod._noise_model_set_phase_damping(ptr, options.phaseDampingRate);
    }
    if (options.t1 !== undefined && options.t2 !== undefined) {
      mod._noise_model_set_thermal(ptr, options.t1, options.t2);
    }
    if (options.gateTime !== undefined) {
      mod._noise_model_set_gate_time(ptr, options.gateTime);
    }
    if (options.readoutError !== undefined) {
      mod._noise_model_set_readout_error(
        ptr, options.readoutError[0], options.readoutError[1],
      );
    }
    return m;
  }

  /** Build a model from device-style figures of merit: T1/T2 in
   *  microseconds, a per-gate error rate, and a readout error
   *  probability (`noise_model_create_realistic`). */
  static async realistic(
    t1Us: number, t2Us: number, gateError: number, readoutErrorProb: number,
    rng?: RandomSource,
  ): Promise<DeviceNoiseModel> {
    const mod = (await getModule()) as unknown as NoiseModule;
    const ptr = mod._noise_model_create_realistic(t1Us, t2Us, gateError, readoutErrorProb);
    if (!ptr) throw new Error('noise_model_create_realistic returned NULL');
    return new DeviceNoiseModel(ptr, mod, rng ?? defaultRng);
  }

  /** Internal: WASM pointer to the underlying `noise_model_t`. */
  _internal_ptr(): number {
    if (this.freed) throw new Error('DeviceNoiseModel disposed');
    return this.ptr;
  }

  /** Enable / disable the model (disabled models are no-ops). */
  setEnabled(enabled: boolean): void {
    this.mod._noise_model_set_enabled(this._internal_ptr(), enabled ? 1 : 0);
  }

  private applyWithDraws(
    stateFn: (rvsPtr: number) => void, numDraws: number,
    randomValues?: readonly number[],
  ): void {
    const mod = this.mod;
    const rvsPtr = mod._malloc(numDraws * 8);
    try {
      for (let i = 0; i < numDraws; i++) {
        mod.HEAPF64[(rvsPtr >> 3) + i] =
          randomValues !== undefined && i < randomValues.length
            ? randomValues[i]
            : this.rng();
      }
      stateFn(rvsPtr);
    } finally {
      mod._free(rvsPtr);
    }
  }

  /** Apply one noise trajectory to `qubit` (call after a single-qubit
   *  gate).  Consumes up to 6 uniform draws. */
  applySingle(state: QuantumState, qubit: number, randomValues?: readonly number[]): void {
    const ptr = this._internal_ptr();
    const { mod, ptr: statePtr } = modAndPtr(state);
    this.applyWithDraws(
      (rvsPtr) => mod._noise_apply_model(statePtr, qubit, ptr, rvsPtr),
      6, randomValues,
    );
  }

  /** Apply one noise trajectory to a two-qubit gate on `qubitA`,
   *  `qubitB`.  Consumes up to 10 uniform draws. */
  applyTwo(
    state: QuantumState, qubitA: number, qubitB: number,
    randomValues?: readonly number[],
  ): void {
    const ptr = this._internal_ptr();
    const { mod, ptr: statePtr } = modAndPtr(state);
    this.applyWithDraws(
      (rvsPtr) => mod._noise_apply_model_two_qubit(statePtr, qubitA, qubitB, ptr, rvsPtr),
      10, randomValues,
    );
  }

  /** Release the underlying C allocation. */
  dispose(): void {
    if (this.freed) return;
    this.mod._noise_model_destroy(this.ptr);
    this.freed = true;
  }
}

File: bindings/rust/moonlab/src/noise.rs
//! Single-qubit Kraus noise channels (C-side since v0.2.1, Rust
//! wrapper since v0.4.8).
//!
//! Wraps `src/quantum/noise.{c,h}` with idiomatic Rust around the
//! per-channel sampler entry points.  Mirrors the v0.2.1 Python
//! `moonlab.noise` surface.
//!
//! Each channel takes `(state, qubit, ..., random_value)`.  The
//! caller supplies the uniform-`[0, 1)` random sample; no
//! `quantum_entropy_ctx_t` is needed because the Kraus selection is
//! decoupled from the sampling primitive.  This makes the channels
//! cheap to use in test benches and deterministic when fed a seeded
//! RNG.
//!
//! ## Quick start
//!
//! ```no_run
//! use moonlab::{QuantumState, noise};
//!
//! // 1% depolarising noise on qubit 0 of a fresh state.
//! let mut state = QuantumState::new(2).unwrap();
//! state.h(0);
//! noise::depolarizing_single(&mut state, 0, /*p=*/0.01, /*r=*/0.42);
//! ```

use crate::error::{QuantumError, Result};
use crate::state::QuantumState;
use moonlab_sys as ffi;

fn check_qubit(state: &QuantumState, q: usize) -> Result<()> {
    if q >= state.num_qubits() {
        return Err(QuantumError::InvalidQubit {
            index: q,
            max: state.num_qubits(),
        });
    }
    Ok(())
}

/// Single-qubit depolarising channel
/// `rho |-> (1 - p) rho + (p / 3) (X rho X + Y rho Y + Z rho Z)`.
/// `random_value` selects which Pauli (if any) to apply.
pub fn depolarizing_single(
    state: &mut QuantumState,
    qubit: usize,
    probability: f64,
    random_value: f64,
) -> Result<()> {
    check_qubit(state, qubit)?;
    unsafe {
        ffi::noise_depolarizing_single(
            state.as_ptr(),
            qubit as i32,
            probability,
            random_value,
        );
    }
    Ok(())
}

/// Two-qubit depolarising channel acting on the pair `(qubit1, qubit2)`.
pub fn depolarizing_two_qubit(
    state: &mut QuantumState,
    qubit1: usize,
    qubit2: usize,
    probability: f64,
    random_value: f64,
) -> Result<()> {
    check_qubit(state, qubit1)?;
    check_qubit(state, qubit2)?;
    unsafe {
        ffi::noise_depolarizing_two_qubit(
            state.as_ptr(),
            qubit1 as i32,
            qubit2 as i32,
            probability,
            random_value,
        );
    }
    Ok(())
}

/// Amplitude-damping (T1) channel with damping parameter `gamma`.
pub fn amplitude_damping(
    state: &mut QuantumState,
    qubit: usize,
    gamma: f64,
    random_value: f64,
) -> Result<()> {
    check_qubit(state, qubit)?;
    unsafe {
        ffi::noise_amplitude_damping(
            state.as_ptr(),
            qubit as i32,
            gamma,
            random_value,
        );
    }
    Ok(())
}

/// Phase-damping (pure-T2) channel with damping parameter `gamma`.
pub fn phase_damping(
    state: &mut QuantumState,
    qubit: usize,
    gamma: f64,
    random_value: f64,
) -> Result<()> {
    check_qubit(state, qubit)?;
    unsafe {
        ffi::noise_phase_damping(
            state.as_ptr(),
            qubit as i32,
            gamma,
            random_value,
        );
    }
    Ok(())
}

/// Pure-dephasing channel with phase variance `sigma`.  `random_phase`
/// is a uniform-`[0, 1)` sample used to draw the random phase angle.
pub fn pure_dephasing(
    state: &mut QuantumState,
    qubit: usize,
    sigma: f64,
    random_phase: f64,
) -> Result<()> {
    check_qubit(state, qubit)?;
    unsafe {
        ffi::noise_pure_dephasing(
            state.as_ptr(),
            qubit as i32,
            sigma,
            random_phase,
        );
    }
    Ok(())
}

/// Bit-flip channel `(1 - p) I + p X`.
pub fn bit_flip(
    state: &mut QuantumState,
    qubit: usize,
    probability: f64,
    random_value: f64,
) -> Result<()> {
    check_qubit(state, qubit)?;
    unsafe {
        ffi::noise_bit_flip(
            state.as_ptr(),
            qubit as i32,
            probability,
            random_value,
        );
    }
    Ok(())
}

/// Phase-flip channel `(1 - p) I + p Z`.
pub fn phase_flip(
    state: &mut QuantumState,
    qubit: usize,
    probability: f64,
    random_value: f64,
) -> Result<()> {
    check_qubit(state, qubit)?;
    unsafe {
        ffi::noise_phase_flip(
            state.as_ptr(),
            qubit as i32,
            probability,
            random_value,
        );
    }
    Ok(())
}

/// Bit-phase-flip channel `(1 - p) I + p Y`.
pub fn bit_phase_flip(
    state: &mut QuantumState,
    qubit: usize,
    probability: f64,
    random_value: f64,
) -> Result<()> {
    check_qubit(state, qubit)?;
    unsafe {
        ffi::noise_bit_phase_flip(
            state.as_ptr(),
            qubit as i32,
            probability,
            random_value,
        );
    }
    Ok(())
}

/// Thermal-relaxation channel combining amplitude-damping at rate `1
/// / t1` and phase-damping at rate `1 / t2` over duration `time`.
/// `random_values` must provide the two uniform-`[0, 1)` samples the
/// underlying combined-Kraus sampler consumes.
pub fn thermal_relaxation(
    state: &mut QuantumState,
    qubit: usize,
    t1: f64,
    t2: f64,
    time: f64,
    random_values: &[f64],
) -> Result<()> {
    check_qubit(state, qubit)?;
    if random_values.len() < 2 {
        return Err(QuantumError::Ffi(format!(
            "thermal_relaxation needs at least 2 random_values, got {}",
            random_values.len()
        )));
    }
    unsafe {
        ffi::noise_thermal_relaxation(
            state.as_ptr(),
            qubit as i32,
            t1,
            t2,
            time,
            random_values.as_ptr(),
        );
    }
    Ok(())
}

/// Classical readout-error model: flip the measurement outcome with
/// probability `error_0_to_1` (when the true outcome is 0) or
/// `error_1_to_0` (when 1).  Returns the noisy outcome.
pub fn readout_error(
    outcome: bool,
    error_0_to_1: f64,
    error_1_to_0: f64,
    random_value: f64,
) -> bool {
    let v = unsafe {
        ffi::noise_readout_error(
            outcome as i32,
            error_0_to_1,
            error_1_to_0,
            random_value,
        )
    };
    v != 0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn depolarizing_single_zero_probability_is_identity() {
        let mut state = QuantumState::new(2).unwrap();
        state.h(0);
        let p_before = state.probabilities();
        depolarizing_single(&mut state, 0, 0.0, 0.5).unwrap();
        let p_after = state.probabilities();
        for (a, b) in p_before.iter().zip(p_after.iter()) {
            assert!((a - b).abs() < 1e-12);
        }
    }

    #[test]
    fn bit_flip_full_probability_flips_qubit() {
        // |0> --bit_flip(p=1)--> |1>; only basis state 1 should
        // have probability 1 afterwards.
        let mut state = QuantumState::new(1).unwrap();
        bit_flip(&mut state, 0, 1.0, 0.0).unwrap();
        let p = state.probabilities();
        assert!(p[0].abs() < 1e-12);
        assert!((p[1] - 1.0).abs() < 1e-12);
    }

    #[test]
    fn phase_flip_full_probability_preserves_basis_probabilities() {
        // Z on |+> goes to |->, but the computational-basis
        // probabilities stay (0.5, 0.5).
        let mut state = QuantumState::new(1).unwrap();
        state.h(0);
        phase_flip(&mut state, 0, 1.0, 0.0).unwrap();
        let p = state.probabilities();
        assert!((p[0] - 0.5).abs() < 1e-12);
        assert!((p[1] - 0.5).abs() < 1e-12);
    }

    #[test]
    fn amplitude_damping_eventually_collapses_excited_state() {
        // |1> with full damping (gamma = 1) -- the excited state
        // decays to |0>.  The C side selects the K_1 (decay) Kraus
        // when random_value < gamma.
        let mut state = QuantumState::new(1).unwrap();
        state.x(0);
        amplitude_damping(&mut state, 0, 1.0, 0.5).unwrap();
        let p = state.probabilities();
        assert!((p[0] - 1.0).abs() < 1e-6);
    }

    #[test]
    fn rejects_out_of_range_qubit() {
        let mut state = QuantumState::new(2).unwrap();
        assert!(depolarizing_single(&mut state, 5, 0.1, 0.5).is_err());
        assert!(bit_flip(&mut state, 5, 0.1, 0.5).is_err());
    }

    #[test]
    fn readout_error_threshold_pins_behaviour() {
        // P(flip | true=0) = 0.1; r=0.05 < 0.1 -> flip, r=0.5 -> keep.
        assert_eq!(readout_error(false, 0.1, 0.1, 0.05), true);
        assert_eq!(readout_error(false, 0.1, 0.1, 0.5), false);
    }

    #[test]
    fn mutual_info_bell_pair_equals_one_bit_per_subsystem() {
        // |Phi+> = (|00> + |11>) / sqrt 2.  I(A:B) = 2 * S(A) = 2 bits.
        let mut state = QuantumState::new(2).unwrap();
        state.h(0).cnot(0, 1);
        let i_ab = state.mutual_information(&[0], &[1]).unwrap();
        assert!(
            (i_ab - 2.0).abs() < 1e-6,
            "I(A:B) = {} on |Phi+>; expected 2 bits",
            i_ab
        );
    }
}
